// @sap/cli-core/config/core.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getGenericOptionsHelp = exports.getDeprecationMessage = exports.isDeprecated = exports.getVersion = exports.getAuthenticationMethods = exports.getSapHelp = exports.getDiscoveryPaths = exports.getDescription = exports.getPackageName = exports.getName = void 0;
/* jscpd:ignore-start */
const constants_1 = require("../constants");
/* jscpd:ignore-end */
const index_1 = require("./index");
const getName = () => (0, index_1.get)()[constants_1.CLI_NAME];
exports.getName = getName;
const getPackageName = () => (0, index_1.get)()[constants_1.CLI_PACKAGE_NAME];
exports.getPackageName = getPackageName;
const getDescription = () => (0, index_1.get)()[constants_1.CLI_DESCRIPTION];
exports.getDescription = getDescription;
const getDiscoveryPaths = () => (0, index_1.get)()[constants_1.CLI_DISCOVERY_PATHS];
exports.getDiscoveryPaths = getDiscoveryPaths;
const getSapHelp = () => (0, index_1.get)()[constants_1.CLI_SAP_HELP];
exports.getSapHelp = getSapHelp;
const getAuthenticationMethods = () => (0, index_1.get)()[constants_1.CLI_SUPPORTED_AUTHENTICATION_METHODS];
exports.getAuthenticationMethods = getAuthenticationMethods;
const getVersion = () => (0, index_1.get)()[constants_1.CLI_VERSION];
exports.getVersion = getVersion;
const isDeprecated = () => (0, index_1.get)()[constants_1.CLI_DEPRECATED] || false;
exports.isDeprecated = isDeprecated;
const getDeprecationMessage = () => (0, index_1.get)()[constants_1.CLI_DEPRECATION_MESSAGE];
exports.getDeprecationMessage = getDeprecationMessage;
const getGenericOptionsHelp = () => (0, index_1.get)()[constants_1.CLI_GENERIC_OPTIONS_HELP];
exports.getGenericOptionsHelp = getGenericOptionsHelp;
