// @sap/cli-core/index.js
"use strict";
/* eslint-disable import/first */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.configureCoreModule = exports.getPackageName = exports.getName = exports.getDescription = exports.getBin = exports.getVersion = exports.AuthenticationMethod = exports.getCommands = exports.configure = exports.run = void 0;
const dotenv = __importStar(require("dotenv"));
const configureLoggers_1 = require("./configureLoggers");
dotenv.config();
(0, configureLoggers_1.configureLoggers)();
const config_1 = require("./config");
/* jscpd:ignore-start */
const constants_1 = require("./constants");
Object.defineProperty(exports, "AuthenticationMethod", { enumerable: true, get: function () { return constants_1.AuthenticationMethod; } });
/* jscpd:ignore-end */
var run_1 = require("./dwc/run");
Object.defineProperty(exports, "run", { enumerable: true, get: function () { return run_1.run; } });
var module_1 = require("./module");
Object.defineProperty(exports, "configure", { enumerable: true, get: function () { return module_1.configure; } });
Object.defineProperty(exports, "getCommands", { enumerable: true, get: function () { return module_1.getCommands; } });
var utils_1 = require("./utils/utils");
Object.defineProperty(exports, "getVersion", { enumerable: true, get: function () { return utils_1.getVersion; } });
Object.defineProperty(exports, "getBin", { enumerable: true, get: function () { return utils_1.getBin; } });
Object.defineProperty(exports, "getDescription", { enumerable: true, get: function () { return utils_1.getDescription; } });
Object.defineProperty(exports, "getName", { enumerable: true, get: function () { return utils_1.getName; } });
Object.defineProperty(exports, "getPackageName", { enumerable: true, get: function () { return utils_1.getPackageName; } });
const configureCoreModule = ({ name, packageName, discoveryPaths, description, sapHelpLink, authenticationMethods = [constants_1.AuthenticationMethod.oauth], version, genericOptionsHelp, deprecated = false, deprecationMessage, }) => {
    (0, config_1.set)({
        [constants_1.CLI_NAME]: name,
        [constants_1.CLI_PACKAGE_NAME]: packageName,
        [constants_1.CLI_DESCRIPTION]: description,
        [constants_1.CLI_DISCOVERY_PATHS]: discoveryPaths,
        [constants_1.CLI_SAP_HELP]: sapHelpLink,
        [constants_1.CLI_SUPPORTED_AUTHENTICATION_METHODS]: authenticationMethods,
        [constants_1.CLI_VERSION]: version,
        [constants_1.CLI_DEPRECATED]: deprecated,
        [constants_1.CLI_DEPRECATION_MESSAGE]: deprecationMessage,
        [constants_1.CLI_GENERIC_OPTIONS_HELP]: genericOptionsHelp,
    });
};
exports.configureCoreModule = configureCoreModule;
