// @sap/cli-core/cache/cache.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFiles = exports.cleanCache = exports.deleteFile = exports.writeFile = exports.readFile = exports.init = void 0;
const fs_extra_1 = __importDefault(require("fs-extra"));
const logger_1 = require("../logger");
const utils_1 = require("./utils");
const getLogger = () => (0, logger_1.get)("cache");
const init = async () => {
    const { trace } = getLogger();
    try {
        trace("initializing cache");
        await fs_extra_1.default.mkdir((0, utils_1.getPath)(), { recursive: true });
    }
    catch (err) {
        trace("error while initializing cache", err.toString());
        throw err;
    }
};
exports.init = init;
const readFile = async (file) => {
    const { trace, debug } = getLogger();
    try {
        trace("reading file %s from cache", file);
        return await fs_extra_1.default.readFile((0, utils_1.getPath)(file), "utf-8");
    }
    catch (err) {
        debug("error while reading file %s from cache", file, err.toString());
        throw err;
    }
};
exports.readFile = readFile;
const writeFile = async (file, data) => {
    const { trace, debug } = getLogger();
    try {
        trace("writing file %s to cache", file);
        await fs_extra_1.default.writeFile((0, utils_1.getPath)(file), data, "utf-8");
    }
    catch (err) {
        debug("error while writing file %s to cache", file, err.toString());
        throw err;
    }
};
exports.writeFile = writeFile;
const deleteFile = async (file) => {
    const { trace, debug } = getLogger();
    try {
        trace("deleting file %s from cache", file);
        await fs_extra_1.default.remove((0, utils_1.getPath)(file));
    }
    catch (err) {
        debug("error while deleting file %s from cache", file, err.toString());
        throw err;
    }
};
exports.deleteFile = deleteFile;
const cleanCache = async () => {
    const { trace, debug } = getLogger();
    try {
        trace("cleaning cache");
        await fs_extra_1.default.remove((0, utils_1.getPath)());
    }
    catch (err) {
        debug("error while cleaning cache", err.toString());
        throw err;
    }
};
exports.cleanCache = cleanCache;
const getFiles = async () => {
    const { trace, debug } = getLogger();
    try {
        trace("reading existing files");
        return await fs_extra_1.default.readdir((0, utils_1.getPath)(), "utf-8");
    }
    catch (err) {
        debug("error while reading existing files", err.toString());
        throw err;
    }
};
exports.getFiles = getFiles;
