// @sap/cli-core/cache/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPath = exports.getFiles = exports.writeFile = exports.deleteFile = exports.cleanCache = exports.init = exports.readFile = void 0;
var cache_1 = require("./cache");
Object.defineProperty(exports, "readFile", { enumerable: true, get: function () { return cache_1.readFile; } });
Object.defineProperty(exports, "init", { enumerable: true, get: function () { return cache_1.init; } });
Object.defineProperty(exports, "cleanCache", { enumerable: true, get: function () { return cache_1.cleanCache; } });
Object.defineProperty(exports, "deleteFile", { enumerable: true, get: function () { return cache_1.deleteFile; } });
Object.defineProperty(exports, "writeFile", { enumerable: true, get: function () { return cache_1.writeFile; } });
Object.defineProperty(exports, "getFiles", { enumerable: true, get: function () { return cache_1.getFiles; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "getPath", { enumerable: true, get: function () { return utils_1.getPath; } });
